src/components/TablePagination/TablePagination.tsx

```diff
@@ -85,94 +45,150 @@ export type Props = PropsWithChildren<{
    * place the pagination component above or below the table?
    */
   position?: "above" | "below";
-}> &
+};
+
+export type ExternalControlProps = BasePaginationProps & {
+  /**
+   * Whether the component will be controlled via external state.
+   */
+  externallyControlled?: true;
+  /**
+   * the total number of items available within the data. This prop is only relevant
+   * and will be required if `externallyControlled` is set to `true`.
+   */
+  totalItems: number;
+  /**
+   * the current page that's showing. This prop is only relevant and will be required
+   * if `externallyControlled` is set to `true`.
+   */
+  currentPage: number;
+  /**
+   * size per page. This prop is only relevant and will be required if
+   * `externallyControlled` is set to `true`.
+   */
+  pageSize: number;
+  /**
+   * callback indicating a page change event to the parent component.
+   * This prop is only relevant and will be required if `externallyControlled` is set
+   * to `true`.
+   */
+  onPageChange: (page: number) => void;
+  /**
+   * callback indicating a page size change event to the parent component.
+   * This prop is only relevant and will be required if `externallyControlled` is set
+   * to `true`.
+   */
+  onPageSizeChange: (pageSize: number) => void;
+};
+
+export type InternalControlProps = BasePaginationProps & {
+  /**
+   * Whether the component will be controlled via external state.
+   */
+  externallyControlled?: false;
+};
+
+export type Props = PropsWithChildren<
+  ExternalControlProps | InternalControlProps
+> &
   HTMLAttributes<HTMLDivElement>;
 
-const TablePagination = ({
-  data,
-  className,
-  itemName = "item",
-  description,
-  position = "above",
-  dataForwardProp = "rows",
-  pageLimits = DEFAULT_PAGE_LIMITS,
-  children,
-  ...divProps
-}: Props) => {
-  const descriptionRef = useRef<HTMLDivElement>(null);
-  const [isSmallScreen, setSmallScreen] = useState(false);
-  const [pageSize, setPageSize] = useState(() => {
+const TablePagination = (props: Props) => {
+  const {
+    data,
+    dataForwardProp = "rows",
+    itemName = "item",
+    className,
+    description,
+    pageLimits = DEFAULT_PAGE_LIMITS,
+    position = "above",
+    externallyControlled,
+    children,
+    ...divProps
+  } = props;
+
+  // Safety check to ensure pageSize is a valid option in
+  // pageLimits if the component is externally controlled
+  if (externallyControlled) {
+    let pageSizeFound = false;
+    for (const limit of pageLimits) {
+      if (limit === Number(props.pageSize)) {
+        pageSizeFound = true;
+        break;
+      }
+    }
+
+    if (!pageSizeFound) {
+      throw new Error(
+        `pageSize must be a valid option in pageLimits, pageLimits is set to [${pageLimits}]`
+      );
+    }
```

Comment:
Shall we check for additional props as well? As in the `externallyControlled` mode, `currentPage`, `onPageChanged`, and other props are mandatory. Maybe we should throw if they are missing?

Reply:
the discriminating union typing for the component will result in Typescript build errors if `externallyControlled` is set to `true` but the required props are not passed. I think defensive code would be redundant in this case.

Reply:
i.e. `ExternalControlProps | InternalControlProps` in the props type definition.